keg_bouncer/model/migration.py

```diff
@@ -21,3 +24,68 @@ def link_user_to_user_groups(op, user_primary_key_column):
 
 def drop_link_from_user_to_user_groups(op):
     op.drop_table(user_user_group_link_table_name)
+
+
+def alembic_create_permission(op, permissions=[{}], groups=set(), bundles=set()):
+    """Bulk create permissions and add them to groups and bundles
+
+    :param permisisons: a list of dictionaries with 'token' and 'description' keys
+    :param groups: a set of group labels, missing groups will be created
+    :param bundles: a set of bundle labels, missing bundles will be created
+    """
+
+    new_permission_tokens = set([perm['token'] for perm in permissions])
+    groups = set(groups)
+    bundles = set(bundles)
+    con = op.get_bind()
+    meta = sa.MetaData()
+    execute = con.execute
+
+    # inspect the tables we need and create a table object for them
+    permissions_table = sa.Table('keg_bouncer_permissions', meta, autoload=True,
+                                 autoload_with=con.engine)
+
+    groups_table = sa.Table('keg_bouncer_user_groups', meta, autoload=True,
+                            autoload_with=con.engine)
+
+    groups_map = sa.Table('keg_bouncer_user_group_permission_map', meta, autoload=True,
+                          autoload_with=con.engine)
+
+    bundles_table = sa.Table('keg_bouncer_permission_bundles', meta, autoload=True,
+                             autoload_with=con.engine)
+
+    bundles_map = sa.Table('keg_bouncer_bundle_permission_map', meta, autoload=True,
+                           autoload_with=con.engine)
```

Comment:
I see what you're saying. That does seem like a possible benefit to doing it this way. However, that only helps if you change the name/type of the column somewhere in the migration *after* this code runs. If your tables have already changed, then you can't use this migration because it hard-codes the column names. In that case, using the ORM would have been better. So it seems to be 6 of one, half-dozen of the other. And using the ORM objects seems more predictable. Ideally, you'd migrate your schema separately from adding new permissions.